### PR TITLE
[SYCL] Fix device selector bug introduced in #12288

### PR DESCRIPTION
**Pull request by aelovikov-intel**
The very last commit in #12288
(https://github.com/intel/llvm/pull/12288/commits/f9a1377f835f86ab71fb6886e07f5e201a0f085b) accidentally removed one condition needed to add the device into allowed devices list. Restore it here.

### CODE REVIEW

**Review by maarquitos14 — COMMENTED** [review]
Is there any test that caught the error? If not, should we add one?

**Review by maarquitos14 — APPROVED** [review]
LGTM. Thanks for adding the test.

File: sycl/test-e2e/DeprecatedFeatures/sycl_device_filter.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  auto devices = sycl::device::get_devices();
+  for (sycl::device d : devices) {
+    std::cout << device_type(d) << " " << d.get_info<dev_info::name>()
+              << std::endl;
+  }
+  assert(!devices.empty());
+  auto expected_type = std::getenv("SYCL_DEVICE_FILTER");
+  assert(std::all_of(devices.begin(), devices.end(), [=](auto d) {
+    return expected_type == device_type(d);
+  }));
```
Comment (aelovikov-intel) [inline sycl/test-e2e/DeprecatedFeatures/sycl_device_filter.cpp:33]:
```suggestion
  assert(std::all_of(devices.begin(), devices.end(),
                     [=](auto d) { return expected_type == device_type(d); }));
```

**Comment by aelovikov-intel** [comment]
CUDA failure looks like some race condition in the CI itself, not a real failure for this PR. Previuos version passed at https://github.com/intel/llvm/actions/runs/7466771596.

**Pull request merged**